torch/csrc/jit/passes/onnx/remove_inplace_ops_for_onnx.cpp

```diff
@@ -19,17 +19,95 @@ namespace {
 const std::set<c10::Symbol> inplace_ops =
     {aten::append, aten::index_put_, aten::pop, aten::insert, aten::Delete};
 
-bool IsInplaceNode(const Node* n) {
-  if (inplace_ops.find(n->kind()) != inplace_ops.end()) {
-    return true;
-  }
-
-  if (n->kind() == Symbol::fromQualString("onnx::Placeholder") &&
-      n->s(attr::name) == "index_put_") {
-    return true;
+// InplaceConverter defines a set of functions that together enables the conversion
+// from prim::GetAttr, prim::SetAttr, and ATen in-place operators to ONNX out-place operators.
+struct InplaceConverter {
+  InplaceConverter(
+      std::shared_ptr<Graph> graph,
+      MutationRemover* mr,
+      Module* model = nullptr)
+      : graph_(std::move(graph)), mr_(mr), module_(model) {}
+
+  void convertMutationForONNX();
+
+ private:
+  void gatherAttrNameInitialValueMap(
+      Block* block,
+      std::unordered_map<std::string, Value*>& attr_name_value_map,
+      std::unordered_map<Node*, std::string>& attr_node_fullname_map);
+  void replaceAttrWithInplaceOps(
+      Block* block,
+      const std::unordered_map<std::string, Value*>& attr_name_value_map,
+      const std::unordered_map<Node*, std::string>& attr_node_fullname_map);
+
+  void convertInplaceOpsAndTrackAlias();
+  void convertInplaceOpsAndTrackAlias(Block* block);
+
+  void correctAliasReferences();
+  void correctAliasReferences(Block* block);
+  void correctAliasReferences(Node* n);
+
+  void convertGetSetAttrToInplaceOps(Block* block);
+
+  // ValueTracker provides apis to record aliases for a single value,
+  // and to retrieve the correct alias of any given value based on the location in the graph it is used.
+  struct ValueTracker {
+    ValueTracker() : graph_(nullptr) {}
+
+    void init(const std::shared_ptr<Graph>& graph);
+    void recordSetValue(Value* old_v, Value* new_v);
+    Value* findAliasForValueAtNode(Value* v, const Node* n) const;
+
+    std::string toString() const;
+
+   private:
+
+    std::shared_ptr<Graph> graph_;
+
+    // Map from aliases to root value.
+    // A single value can have multiple aliases throughout the graph,
+    // created by inplace operators, and preserved through loop carried
+    // input/output. For each such value, its first occurance will be set as
+    // root value.
+    std::unordered_map<Value*, Value*> alias_to_value_;
+
+    // Sort the alias based on their order in graph.
+    // A tie can happen when two aliases belong to different blocks, while having the same ancestor node.
+    // The unique id is used as tie breaker.
+    struct aliasComp {
+      bool operator()(const Value* a, const Value* b) const {
+        auto* n_a = a->node();
+        auto* n_b = b->node();
+        if (n_a == n_b) {
+          return false;
+        }
+        auto a_b = n_a->isBefore(n_b);
+        auto b_a = n_b->isBefore(n_a);
+        if (a_b == b_a) {
+          return a->unique() < b->unique();
```

Comment:
Why is the unique id used as a tiebreaker?

Reply:
unique id is guaranteed to be different for different `Value` obj. Ties must be broken, otherwise the two aliases will be considered equal to each other, and then only one of them gets stored into the map.